spec/registry.md

```diff
@@ -359,6 +359,14 @@ the `:string` selector performs as described below.
 
 The `:string` function returns the string value of the resolved value of the _operand_.
 
+#### Composition
+
+When an _operand_ or an _option_ value uses a _variable_ with a _declaration_
+using an _expression_ with a `:string` _annotation_,
+its resolved value contains only the string value of the _operand_ of the annotated _expression_,
+together with its resolved locale and directionality,
+and no _option_ values.
```

Comment:
I think this could be improved. The first part of the sentence is really hard to read. Also, I think that we should use normative language and we should permit some implementation-specific things to occur.

```suggestion
The _resolved value_ of an _expression_ with a `:string` _annotation_ consists of
the string value of the _operand_ of the annotated _expression_.
It MUST also include the `locale`, `direction`, and `id`, if they exist, from the formatting context.
It MAY include implementation-defined attributes or option values.
```

Reply:
```suggestion
The _resolved value_ of an _expression_ with a `:string` _annotation_ consists of:
- the string value of the _operand_ of the annotated _expression_
- if they exist, the `locale`, `direction`, and `id` from the formatting context
- optionally, implementation-defined attributes or option values
```
Reason: it is strange to say that "X consists of Y", which implies that X consists of exactly that, no more, no less. So reworded to list all the things it could consist of, marking the ones that are optional.

Reply:
I am in general fine with this sort of an approach to defining _resolved value_, but it'll mean that we need to do so more deeply than this small change I'm trying to propose, because we need to account for what happens here:

```
.input {$x :string}
.match {$x}
foo {{{$x} is foo}}
* {{{$x} isn't foo, but as a number it's {$x :number}}}
```

Where do we place the matching and formatting behaviour for `$x` if its _resolved value_ only consists of data fields? Even though those are relatively simple for `:string`, our general solution will need to work for all sorts of functions, and provide a way for defining how their formatting, matching, and composition works when all you have is a _resolved value_.

As I mentioned most recently in https://github.com/unicode-org/message-format-wg/issues/515#issuecomment-1797871157, such a resolved value is almost certainly most easily defined as an object with different methods for its different use cases.

However, defining the spec in such a manner has received a significant amount of pushback, and so it's currently done by only defining the observable parts, i.e. in this case the resolved value of a _variable_ referring to a `:string` annotated value when that _variable_ is used as an operand or an option value.

If we do want to define _resolved values_ more explicitly, then we need to go all in on that approach, and start that definition from the base class definition of something like a ResolvedValue with abstract methods, which would allow for a ResolvedString child class to be defined as the _resolved value_ of a `:string` expression, together with the behaviour of at least its `getValue`, `getOptions`, `formatToString`, and `matchKeys` methods.

Or we can stick to our current approach of not explaining the internals, and directly describing the observable parts of that behaviour.

Reply:
To get an idea for what such a spec could look like, please take a look at #198, my first attempt at doing so. Our language has changed in the interim, and that exact "Formattable" API doesn't match our current needs (e.g. it's missing `getOptions`, and we dropped formatted parts), but something like that could be a way for us to define a proper _resolved value_.

Reply:
You're not wrong, but what I would say is... the "resolved value" is the value of the operand (the input), not the value of the output (the formatted string or Formattable object in the case of a formatter or the sorted options list in the case of a selector).

Our existing functions do not alter the operand, so their "resolved value" is whatever the function decides the operand value is. Here's an example that might demonstrate a function with a 'resolved value' side-effect:

```
.input {$x :uppercase @locale=fr}
.match {$x :string}
foo {{Never matches}}
FOO {{{$x} matches if $x is any permutation of foo/Foo/FOO/etc}}
* {{}}
```